---
title: 'Patel: covariates and batch effects'
author: "Davide Risso"
date: "2/23/2017"
output: 
  html_document: 
    fig_height: 7
    fig_width: 7
    toc: yes
    code_folding: hide
    toc_float: yes
---

```{r options, echo=FALSE, results="hide",mesasge=FALSE, error=FALSE, include=FALSE, autodep=TRUE}
knitr::opts_chunk$set(fig.align="center", cache=TRUE, error=FALSE, message=FALSE, warning=TRUE)
library(SummarizedExperiment)
library(scRNAseq)
library(zinbwave)
library(ggplot2)
library(magrittr)
library(matrixStats)
library(Rtsne)
library(scone)
library(RColorBrewer)
library(cluster)
library(parallel)
library(EDASeq)
library(digest)
library(edgeR)
library(DESeq2)

theme_set(theme_bw()+theme(
  axis.title = element_text(size=16),
  plot.title=element_text(hjust=0.5, size=20))
  )

source('../R/utilityFunctions.R')

set.seed(8283)
options(getClass.msg=FALSE) #get rid of annoying messages about cache until fixed internally in R
```

Here, I will explore the relation between the expression values and some covariates provided with the data.

```{r datain}

counts <- read.table("patel/glioblastoma_raw_rnaseq_SCandbulk_counts_withannots.txt", header=TRUE, stringsAsFactors = FALSE)
info <- as.matrix(counts)[1,-(1:3)]

# Drops "X" rownames column and sample info row
counts <- counts[-1,-1]

gene_symbols <- counts[,1]
ensembl_ids <- counts[,2]

sample_names <- colnames(counts)[-(1:2)]

# Drops gene symbol and ID columns
all.counts <- counts[,-(1:2)]
rownames(all.counts) <- ensembl_ids

metadata <- read.table("patel/SraRunTable.txt", sep=',', stringsAsFactors = FALSE, header=TRUE)
rownames(metadata)<-metadata$Run
metadata <- metadata[sample_names,]

batch <- stringr::str_split_fixed(info, "_", 2)[,2]
batch <- stringr::str_split_fixed(batch, "_", 2)[,1]

keep <- which(grepl("^Single cell", metadata$source_name) &
                grepl("MGH", metadata$patient_id) &
		metadata$subtype != "")
		#metadata$subtype != "" & metadata$subtype != "None")
		
metadata <- metadata[keep,]

batch <- as.factor(batch[keep])
all.counts <- all.counts[,keep]
all.counts <- as.matrix(all.counts)
class(all.counts) <- "numeric"

stopifnot(all(rownames(metadata)==colnames(all.counts)))

col1 <- brewer.pal(9, "Set1")
col2 <- c(brewer.pal(8, "Set2"), brewer.pal(8, "Set3"), brewer.pal(8, "Set1"))

detection_rate <- colSums(all.counts>0)
coverage <- colSums(all.counts)

qc <- cbind(detection_rate, coverage)

level1 <- as.factor(metadata$patient_id)
level2 <- as.factor(metadata$subtype)
```

# Comparison of methods

```{r normalization}
filter <- rowSums(all.counts>50)>=50
raw <- all.counts[filter,]

totalcount = function (ei)
{
  sums = colSums(ei)
  eo = t(t(ei)*mean(sums)/sums)
  return(eo)
}

tc <- totalcount(raw)
fq <- FQT_FN(raw)
tmm <- TMM_FN(raw)
deseq <- DESEQ_FN(raw)
```

## PCA

```{r pca}
library(rARPACK)
fastpca <- function(expr, scale=FALSE) {
  svd_raw <- svds(scale(t(expr), center=TRUE, scale=scale), k=3, nu=3, nv=0)
  pc_raw <- svd_raw$u %*% diag(svd_raw$d[1:3])
  return(pc_raw)
}

vars <- rowVars(log1p(tc))
names(vars) <- rownames(tc)
vars <- sort(vars, decreasing = TRUE)
vargenes <- names(vars)[1:1000]

pc_raw <- fastpca(log1p(raw[vargenes,]))
pc_tc <- fastpca(log1p(tc[vargenes,]))
pc_fq <- fastpca(log1p(fq[vargenes,]))
pc_tmm <- fastpca(log1p(tmm[vargenes,]))
#pc_deseq <- fastpca(log1p(deseq))

colMerged <- col1[level1]
plot(pc_raw, col=colMerged, pch=20, main="PCA RAW")
legend("bottomleft", levels(level1), fill=col1)

plot(pc_tc, col=colMerged, pch=20, main="PCA TC")
plot(pc_tmm, col=colMerged, pch=20, main="PCA TMM")
plot(pc_fq, col=colMerged, pch=20, main="PCA FQ")
```

## ZIFA

```{r zifa}

zifa_raw <- wrapRzifa(log1p(raw[vargenes,]), k=2)
pairs(zifa_raw, pch=19, col=colMerged, main="ZIFA Raw")

zifa_tc <- wrapRzifa(log1p(tc[vargenes,]), k=2)
pairs(zifa_tc, pch=19, col=colMerged, main="ZIFA TC")

zifa_tmm <- wrapRzifa(log1p(tmm[vargenes,]), k=2)
pairs(zifa_tmm, pch=19, col=colMerged, main="ZIFA TMM")

zifa_fq <- wrapRzifa(log1p(fq[vargenes,]), k=2)
pairs(zifa_fq, pch=19, col=colMerged, main="ZIFA FQ")
```

## ZINB

```{r zinb}


raw <- as.matrix(raw)
zinb <- fitZinbAndCache(raw[vargenes,], K = 2, epsilon=1e3)
plot(zinb@W, pch=19, col=colMerged, main="ZINB")
# To show MGH26 and MGH26-2 as in figure 5e
plot(zinb@W, pch=19, col=batch, main="ZINB")
```

## Add batches in ZINB model

Not sure if this was intended, but the subset considered comprises cells classified as 'Cla', 'Mes', 'Neu', and 'Pro' (i.e. all have single classifications). The batch is included as a sample-level covariate.

```{r zinb_batch}

subset <- which(level2 %in% levels(level2)[c(1, 3, 4, 8)])

vars <- rowVars(log1p(tc[,subset]))
names(vars) <- rownames(tc)
vars <- sort(vars, decreasing = TRUE)
vargenes2 <- names(vars)[1:1000]

zinb_batch <- fitZinbAndCache(raw[vargenes2, subset], K = 2, X=model.matrix(~batch[subset]), epsilon=1e3)
# 5 colours for tumor type
plot(zinb_batch@W, pch=19, col=colMerged[subset], main="ZINB")
# 4 colours for the four tumor types
plot(zinb_batch@W, pch=19, col=col2[droplevels(level2[subset])], main="ZINB")
```

```{r zinb_subset}

subset <- which(level1=="MGH26")

vars <- rowVars(log1p(tc[,subset]))
names(vars) <- rownames(tc)
vars <- sort(vars, decreasing = TRUE)
vargenes2 <- names(vars)[1:1000]

# Fits ZW to the (filtered) MGH26 data without batch as a covariate
zinb_sub1 <- fitZinbAndCache(raw[vargenes2, subset], K = 2, epsilon=1e3)
# No batch
plot(zinb_sub1@W, pch=19, col=col1[droplevels(batch[subset])], main="ZINB")
# Tumour as colour
plot(zinb_sub1@W, pch=19, col=col2[droplevels(level2[subset])], main="ZINB")

# Fits ZW to the (filtered) MGH26 data with batch as a covariate
zinb_sub2 <- fitZinbAndCache(raw[vargenes2, subset], K = 2, X=model.matrix(~batch[subset]), epsilon=1e3, sampleCovLabel="batch")
# With batch
plot(zinb_sub2@W, pch=19, col=col1[droplevels(batch[subset])], main="ZINB")
# Tumor subtype as colour
plot(zinb_sub2@W, pch=19, col=col2[droplevels(level2[subset])], main="ZINB")

# Tumour subtype as a covariate
zinb_sub3 <- fitZinbAndCache(raw[vargenes2, subset], K = 2, X=model.matrix(~factor(level2[subset])), epsilon=1e3, sampleCovLabel="tumour_subtype")
plot(zinb_sub3@W, pch=19, col=batch[subset])
mm = model.matrix(~subtype+batch, data.frame(subtype=factor(level2[subset]), batch=factor(batch[subset])))
# Tumour subtype and batch as covariates
zinb_sub4 <- fitZinbAndCache(raw[vargenes2, subset], K = 2, X=mm, epsilon=1e3, sampleCovLabel="tumour_subtype_batch")
plot(zinb_sub4@W, pch=19, col=batch[subset])
# Detection rate
zinb_sub5 <- fitZinbAndCache(raw[vargenes2, subset], K = 2, X=model.matrix(~detection_rate[subset]), epsilon=1e3, sampleCovLabel="subset_detection_rate")
```
Detection rate was taken as the number of genes/features with one or more reads (no division).

```{r zinb_detection_rate}

zinb_det <- fitZinbAndCache(raw[vargenes,], K = 2, X=model.matrix(~detection_rate), epsilon=1e3, sampleCovLabel = "detRate")
plot(zinb_det@W, pch=19, col=colMerged, main="ZINB with det. rate as covariate", xlab="W1", ylab="W2")
# To show MGH26 and MGH26-2 as in figure 5e
plot(zinb_det@W, pch=19, col=batch, main="ZINB with detection rate as covariate")
plot(zinb@W, pch=19, col=colMerged, main="ZINB")
```
```{r presPlots}
setwd("/home/tom/diss/zinb_analysis/real_data/patel_plots")

pl1<-ggplot(data=data.frame(zinb@W, Batch=batch), aes(x=X1, y=X2, col=Batch))+geom_point()+xlab("W1")+ylab("W2")+ggtitle("Intercept-only model")+theme(aspect.ratio=1)
pl1

ggsave(pl1, file="interceptOnly_zw.pdf", width=5, height=5)

pl2<-ggplot(data=data.frame(zinb_det@W, Batch=batch), aes(x=X1, y=X2, col=Batch))+geom_point()+xlab("W1")+ylab("W2")+ggtitle("Detection rate as a covariate")+theme(aspect.ratio=1)
pl2

ggsave(pl2, file="detRate_zw.pdf", width=5, height=5)

pl3<-ggplot(data=data.frame(zinb_sub1@W, Batch=factor(batch[subset])), aes(x=X1, y=X2, col=Batch))+geom_point()+xlab("W1")+ylab("W2")+ggtitle("Intercept-only model")+theme(aspect.ratio=1, legend.position="None")
pl3

ggsave(pl3, file="mgh26_interceptOnly_zw.pdf", width=5, height=5)

pl4<-ggplot(data=data.frame(zinb_sub2@W, Batch=factor(batch[subset])), aes(x=X1, y=X2, col=Batch))+geom_point()+xlab("W1")+ylab("W2")+ggtitle("Batch as covariate")+theme(aspect.ratio=1, legend.position="None")
pl4

ggsave(pl4, file="mgh26_batch_zw.pdf", width=5, height=5)

pl5<-ggplot(data=data.frame(zinb_sub3@W, Batch=factor(batch[subset])), aes(x=X1, y=X2, col=Batch))+geom_point()+xlab("W1")+ylab("W2")+ggtitle("Tumour subtype as covariate")+theme(aspect.ratio=1, legend.position="None")
pl5

ggsave(pl5, file="mgh26_tumour_zw.pdf", width=5, height=5)

pl6<-ggplot(data=data.frame(zinb_sub4@W, Batch=factor(batch[subset])), aes(x=X1, y=X2, col=Batch))+geom_point()+xlab("W1")+ylab("W2")+ggtitle("Batch and tumour subtype as covariates")+theme(aspect.ratio=1)
pl6

ggsave(pl6, file="mgh26_tumour_batch_zw.pdf", width=5, height=5)

pl7<-ggplot(data=data.frame(zinb_sub5@W, Batch=factor(batch[subset])), aes(x=X1, y=X2, col=Batch))+geom_point()+xlab("W1")+ylab("W2")+ggtitle("Detection rate as covariate")+theme(aspect.ratio=1, legend.position="None")
pl7

ggsave(pl7, file="mgh26_detRate_zw.pdf", width=5, height=5)
```

# Gene-level covariates

```{r gc_content, eval=FALSE}
counts <- counts[bm[,1],]
gene_info = getGeneLengthAndGCContent(as.character(bm[,2]), "mm10", mode="org.db")
rownames(gene_info) = bm[,1]
gene_info = na.omit(gene_info)
```

```{r write_out, dependson=c("zinb", "zifa", "pca", "zinb_batch", "zinb_detection_rate")}
save(zinb, zinb_batch, zinb_det, zifa_fq, zifa_tmm, zifa_tc, zifa_raw, pc_tmm, pc_fq, pc_tc, pc_raw, file="patel_covariates.rda")
```
