---
title: 'Patel: covariates and batch effects'
author: "Davide Risso"
date: "2/23/2017"
output: 
  html_document: 
    fig_height: 7
    fig_width: 7
    toc: yes
    code_folding: hide
    toc_float: yes
---

```{r options, echo=FALSE, results="hide",mesasge=FALSE, error=FALSE, include=FALSE, autodep=TRUE}
knitr::opts_chunk$set(fig.align="center", cache=TRUE, error=FALSE, message=FALSE, warning=TRUE)
library(SummarizedExperiment)
library(scRNAseq)
library(zinbwave)
library(ggplot2)
library(magrittr)
library(matrixStats)
library(Rtsne)
library(scone)
library(RColorBrewer)
library(cluster)
library(parallel)
library(EDASeq)
library(digest)
library(edgeR)
library(DESeq2)

set.seed(8283)
options(getClass.msg=FALSE) #get rid of annoying messages about cache until fixed internally in R

ncores <- 7
```

Here, I will explore the relation between the expression values and some covariates provided with the data.

```{r datain}
# Adapting Risso's code from the Rmd file for processing their own Patel counts
counts <- read.table("patel/glioblastoma_raw_rnaseq_SCandbulk_counts_withannots.txt", header=TRUE, stringsAsFactors = FALSE)
info <- as.matrix(counts)[1,-(1:3)]

# Drops "X" rownames column and sample info row
counts <- counts[-1,-1]

gene_symbols <- counts[,1]
ensembl_ids <- counts[,2]

sample_names <- colnames(counts)[-(1:2)]

# Drops gene symbol and ID columns
all.counts <- counts[,-(1:2)]
rownames(all.counts) <- ensembl_ids

# Risso's code
# metadata <- read.table("patel/SraRunTable.txt", sep='\t', stringsAsFactors = FALSE, header=TRUE, row.names=5, na.strings = "<not provided>")

metadata <- read.table("patel/SraRunTable.txt", sep=',', stringsAsFactors = FALSE, header=TRUE)
rownames(metadata)<-metadata$Run
metadata <- metadata[sample_names,]

batch <- stringr::str_split_fixed(info, "_", 2)[,2]
batch <- stringr::str_split_fixed(batch, "_", 2)[,1]


# Subtype filtering is odd, what was the intention? Apparently there is a difference between an empty string and 'None'
# May need to revisit this
keep <- which(grepl("^Single cell", metadata$source_name) &
                grepl("MGH", metadata$patient_id) &
		metadata$subtype != "")
		#metadata$subtype != "" & metadata$subtype != "None")
		
metadata <- metadata[keep,]

batch <- as.factor(batch[keep])
all.counts <- all.counts[,keep]
all.counts <- as.matrix(all.counts)
class(all.counts) <- "numeric"

stopifnot(all(rownames(metadata)==colnames(all.counts)))

col1 <- brewer.pal(9, "Set1")
col2 <- c(brewer.pal(8, "Set2"), brewer.pal(8, "Set3"), brewer.pal(8, "Set1"))

detection_rate <- colSums(all.counts>0)
coverage <- colSums(all.counts)

qc <- cbind(detection_rate, coverage)

level1 <- as.factor(metadata$patient_id)
level2 <- as.factor(metadata$subtype)
```

# Comparison of methods

```{r normalization}
filter <- rowSums(all.counts>50)>=50
raw <- all.counts[filter,]

totalcount = function (ei)
{
  sums = colSums(ei)
  eo = t(t(ei)*mean(sums)/sums)
  return(eo)
}

tc <- totalcount(raw)
fq <- FQT_FN(raw)
tmm <- TMM_FN(raw)
deseq <- DESEQ_FN(raw)
```

## PCA

```{r pca}
library(rARPACK)
fastpca <- function(expr, scale=FALSE) {
  svd_raw <- svds(scale(t(expr), center=TRUE, scale=scale), k=3, nu=3, nv=0)
  pc_raw <- svd_raw$u %*% diag(svd_raw$d[1:3])
  return(pc_raw)
}

vars <- rowVars(log1p(tc))
names(vars) <- rownames(tc)
vars <- sort(vars, decreasing = TRUE)
vargenes <- names(vars)[1:1000]

pc_raw <- fastpca(log1p(raw[vargenes,]))
pc_tc <- fastpca(log1p(tc[vargenes,]))
pc_fq <- fastpca(log1p(fq[vargenes,]))
pc_tmm <- fastpca(log1p(tmm[vargenes,]))
#pc_deseq <- fastpca(log1p(deseq))

colMerged <- col1[level1]
plot(pc_raw, col=colMerged, pch=20, main="PCA RAW")
legend("bottomleft", levels(level1), fill=col1)

plot(pc_tc, col=colMerged, pch=20, main="PCA TC")
plot(pc_tmm, col=colMerged, pch=20, main="PCA TMM")
plot(pc_fq, col=colMerged, pch=20, main="PCA FQ")
```

## ZIFA

```{r zifa}
wrapRzifa <- function(Y, block = TRUE, k=2, cacheDirPath="zifaCache"){
  # wrapper R function for ZIFA.
  # md5 hashing and temporary files are used not to re-run zifa 
  # if it has already be run on this computer.
  d = digest(Y, "md5")

  cachePath = paste0(cacheDirPath, '/', d)

  write.csv(Y, paste0(cachePath, '.csv'))
  
  if (!file.exists(paste0(cachePath, "_", k, '_zifa.csv'))){
    print('run ZIFA')
    bb = ifelse(block, '-b ', '')
    cmd = sprintf('python run_zifa.py -d %d %s%s.csv %s_%d_zifa.csv', k, bb, cachePath, cachePath, k)
    system(cmd)
  }
  read.csv(sprintf("%s_%d_zifa.csv", cachePath, k), header=FALSE)

}

setwd("/home/tom/diss/zinb_analysis/real_data")

zifa_raw <- wrapRzifa(log1p(raw[vargenes,]), k=2)
pairs(zifa_raw, pch=19, col=colMerged, main="ZIFA Raw")

zifa_tc <- wrapRzifa(log1p(tc[vargenes,]), k=2)
pairs(zifa_tc, pch=19, col=colMerged, main="ZIFA TC")

zifa_tmm <- wrapRzifa(log1p(tmm[vargenes,]), k=2)
pairs(zifa_tmm, pch=19, col=colMerged, main="ZIFA TMM")

zifa_fq <- wrapRzifa(log1p(fq[vargenes,]), k=2)
pairs(zifa_fq, pch=19, col=colMerged, main="ZIFA FQ")
```

## ZINB

```{r zinb}

setwd("/home/tom/diss/zinb_analysis/real_data")

# Defaulting to common dispersion here as that is true of zinbFit
fitZinbAndCache<-function(countMatrix, cacheDirPath="zinbCache", K=2, epsilon=1000, commonDispersion=T) {
  # fit zinb (if you already fitted zinb, it is cached)
  d = digest(countMatrix, "md5")
 
  cachePath = paste0(cacheDirPath, '/', d)
 
  fileZinb = sprintf("%s_%d_%d_%d_zinb.rda", cachePath, K, epsilon, commonDispersion)
 
  if (!file.exists(fileZinb)){
    print('run ZINB')
 
    zinb <- zinbFit(countMatrix, K = K, commondispersion = commonDispersion, epsilon = epsilon)
    save(zinb, file = fileZinb)
  }else{
    load(fileZinb)
  }
  return(zinb)
}

raw <- as.matrix(raw)
zinb <- fitZinbAndCache(raw[vargenes,], K = 2, epsilon=1e3)
plot(zinb@W, pch=19, col=colMerged, main="ZINB")
```

## Add batches in ZINB model

```{r zinb_batch}
setwd("/home/tom/diss/zinb_analsyis/real_data")
# Different signature for design matrix
fitZinbAndCache<-function(countMatrix, X, cacheDirPath="zinbCache", K=2, epsilon=1000, commonDispersion=T) {
  # fit zinb (if you already fitted zinb, it is cached)
  d = digest(countMatrix, "md5")
 
  cachePath = paste0(cacheDirPath, '/', d)
 
  fileZinb = sprintf("%s_%d_%d_%d_zinb.rda", cachePath, K, epsilon, commonDispersion)
 
  if (!file.exists(fileZinb)){
    print('run ZINB')
 
    zinb <- zinbFit(countMatrix, X=X, K = K, commondispersion = commonDispersion, epsilon = epsilon)
    save(zinb, file = fileZinb)
  }else{
    load(fileZinb)
  }
  return(zinb)
}


subset <- which(level2 %in% levels(level2)[c(1, 3, 4, 8)])

vars <- rowVars(log1p(tc[,subset]))
names(vars) <- rownames(tc)
vars <- sort(vars, decreasing = TRUE)
vargenes2 <- names(vars)[1:1000]

zinb_batch <- fitZinbAndCache(raw[vargenes2, subset], K = 2, X=model.matrix(~batch[subset]), epsilon=1e3)
plot(zinb_batch@W, pch=19, col=colMerged[subset], main="ZINB")
plot(zinb_batch@W, pch=19, col=col2[droplevels(level2[subset])], main="ZINB")
```

```{r zinb_subset}
setwd("/home/tom/diss/zinb_analysis/real_data")

subset <- which(level1=="MGH26")

vars <- rowVars(log1p(tc[,subset]))
names(vars) <- rownames(tc)
vars <- sort(vars, decreasing = TRUE)
vargenes2 <- names(vars)[1:1000]

zinb_sub1 <- fitZinbAndCache(raw[vargenes2, subset], K = 2, epsilon=1e3)
plot(zinb_sub1@W, pch=19, col=col1[droplevels(batch[subset])], main="ZINB")
plot(zinb_sub1@W, pch=19, col=col2[droplevels(level2[subset])], main="ZINB")

zinb_sub2 <- fitZinbAndCache(raw[vargenes2, subset], K = 2, X=model.matrix(~batch[subset]), epsilon=1e3)
plot(zinb_sub2@W, pch=19, col=col1[droplevels(batch[subset])], main="ZINB")
plot(zinb_sub2@W, pch=19, col=col2[droplevels(level2[subset])], main="ZINB")
```

# Gene-level covariates

```{r gc_content, eval=FALSE}
counts <- counts[bm[,1],]
gene_info = getGeneLengthAndGCContent(as.character(bm[,2]), "mm10", mode="org.db")
rownames(gene_info) = bm[,1]
gene_info = na.omit(gene_info)
```

```{r write_out, dependson=c("zinb", "zifa", "pca", "zinb_batch")}
save(zinb, zinb_batch, zifa_fq, zifa_tmm, zifa_tc, zifa_raw, pc_tmm, pc_fq, pc_tc, pc_raw, file="patel_covariates.rda")
```
